vnext/Microsoft.ReactNative/ReactPropertyBag.idl

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.ReactNative {
+
+  // The delegate is used to create property value on-demand.
+  [webhosthidden]
+  delegate Object ReactCreatePropertyValue();
+
+  // Namespace for the property name.
+  // Use ReactPropertyNameTable.GetNamespace to get atomic property namespace for a string.
+  [webhosthidden]
+  interface IReactPropertyNamespace {
+    // Get String name representation of the property namespace.
+    String NamespaceName { get; };
+  }
+
+  // Name of a property.
+  // Use ReactPropertyNameTable.GetName to get atomic property name for a string.
+  [webhosthidden]
+  interface IReactPropertyName {
+    // The local property name String in context of the property namespace.
+    String LocalName { get; };
+
+    // The namespace the property name is defined in.
+    IReactPropertyNamespace Namespace { get; };
+  }
+
+  // The IPropertyBag provides a thread-safe property storage.
+  // Properties are identified by a propertyName.
+  // have their own set of properties not visible to other components.
+  // It is expected that there will be no direct use of this interface.
+  // All usage should happen through a strongly typed accessors.
+  [webhosthidden]
+  runtimeclass ReactPropertyBag {
+    // Construct new PropertyNameTable.
+    ReactPropertyBag();
+
+    // Return a global namespace that an empty name.
+    static IReactPropertyNamespace GlobalNamespace { get; };
+
+    // Get an atomic IReactPropertyNamespace for a provided namespaceName.
+    // The valid namespace name must be at least one character long.
+    // Otherwise, the function crashes.
+    // Consider to use module name as a namespace for module-specific properties.
+    static IReactPropertyNamespace GetNamespace(String namespaceName);
+
+    // Get atomic IReactPropertyName for the provided property name and property namespace.
+    // If ns is null, then it uses IReactPropertyNamespace.GlobalNamespace.
+    // The valid property name must be at least one character long.
+    // Otherwise, the function crashes.
+    static IReactPropertyName GetName(IReactPropertyNamespace ns, String localName);
+
+    // Get property value for the propertyName. It returns null if the property does not exist.
+    Object GetProperty(IReactPropertyName propertyName);
+
+    // Get property value for the propertyName.
+    // If the property does not exist, then create it by calling createValue delegate.
+    // The function may return null if the createValue returns null when called.
+    Object GetOrCreateProperty(IReactPropertyName propertyName, ReactCreatePropertyValue createValue);
```

Comment:
>Object [](start = 4, length = 6)

Is there a way to tell whether the property got created or if it already existed?

Reply:
No, there is no way. One can rely on seeing that the creator is called, but it is only true if we won the race. If some other thread assigned the new value, our newly value is thrown away and we pickup theirs.

Reply:
there is a point we should make with this design, IDK if in the comments or the interface definition - that the creator should have no side effects beyond creating the object, because the object that it creates might not end up being used (because we lost the race).

Reply:
I have added it to the comment. It is a quite common practice for concurrent data structures. E.g. see the https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netcore-3.1#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__0_System_Func__0__1__